File: tools/build-standalone.mjs
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

if (process.argv.length !== 3) {
  console.error('Usage: node tools/build-standalone.mjs <output.html>');
  console.error('Exactly one output path is required.');
  process.exit(1);
}

const root = fileURLToPath(new URL('..', import.meta.url));
const output = resolve(process.argv[2]);

function replaceExactlyOnce(source, pattern, replacement, label) {
  let count = 0;
  const result = source.replace(pattern, (...args) => {
    count += 1;
    return typeof replacement === 'function' ? replacement(...args) : replacement;
  });
  if (count !== 1) {
    throw new Error(`Expected exactly one ${label}; found ${count}.`);
  }
  return result;
}

function removeExactlyOnce(source, pattern, label) {
  return replaceExactlyOnce(source, pattern, '', label);
}

function escapeInlineScript(source) {
  return source.replaceAll('</script', '<\\/script');
}

let html = await readFile(resolve(root, 'index.html'), 'utf8');
const css = await readFile(resolve(root, 'styles.css'), 'utf8');
const i18n = escapeInlineScript(await readFile(resolve(root, 'i18n.js'), 'utf8'));
const icon = (await readFile(resolve(root, 'icon-192.png'))).toString('base64');
const gameSource = await readFile(resolve(root, 'game.js'), 'utf8');
const standaloneGame = replaceExactlyOnce(
  gameSource,
  /^[ \t]*if \(['"]serviceWorker['"] in navigator && location\.protocol !== ['"]file:['"]\) \{\r?\n[ \t]*navigator\.serviceWorker\.register\(['"]\.\/sw\.js['"]\)\.catch\(\(\) => \{\}\);\r?\n[ \t]*\}\r?\n/gm,
  '',
  'service worker registration block in game.js',
);
const game = escapeInlineScript(standaloneGame);

html = removeExactlyOnce(
  html,
  /^[ \t]*<link\b(?=[^>]*\brel=["']manifest["'])[^>]*>[ \t]*\r?\n?/gim,
  'manifest link',
);
html = replaceExactlyOnce(
  html,
  /^[ \t]*<link\b(?=[^>]*\brel=["']icon["'])[^>]*>[ \t]*\r?\n?/gim,
  `  <link rel="icon" href="data:image/png;base64,${icon}">\n`,
  'icon link',
);
html = removeExactlyOnce(
  html,
  /^[ \t]*<link\b(?=[^>]*\brel=["']apple-touch-icon["'])[^>]*>[ \t]*\r?\n?/gim,
  'apple-touch-icon link',
);
html = replaceExactlyOnce(
  html,
  /^[ \t]*<link\b(?=[^>]*\brel=["']stylesheet["'])(?=[^>]*\bhref=["'](?:\.\/)?styles\.css["'])[^>]*>[ \t]*\r?\n?/gim,
  `  <style>\n${css}\n  </style>\n`,
  'styles.css link',
);
html = replaceExactlyOnce(
  html,
  /^[ \t]*<script\b(?=[^>]*\bsrc=["'](?:\.\/)?i18n\.js["'])[^>]*>\s*<\/script>[ \t]*\r?\n?/gim,
  `  <script>\n${i18n}\n  </script>\n`,
  'i18n.js script',
);
html = replaceExactlyOnce(
  html,
  /^[ \t]*<script\b(?=[^>]*\bsrc=["'](?:\.\/)?game\.js["'])[^>]*>\s*<\/script>[ \t]*\r?\n?/gim,
  `  <script>\n${game}\n  </script>\n`,
  'game.js script',
);

if (/<script\b[^>]*\bsrc\s*=|<link\b[^>]*\bhref\s*=\s*["'](?!data:)[^"']+|<link\b[^>]*\bimagesrcset\s*=|<(?:img|audio|video|source)\b[^>]*\bsrc\s*=/i.test(html)) {
  throw new Error('Standalone output still contains an external resource reference.');
}

await mkdir(dirname(output), { recursive: true });
await writeFile(output, html, 'utf8');
console.log(output);

File: tools/build_standalone.py
"""Build a self-contained PULSE LINK HTML file with no external runtime assets."""

from __future__ import annotations

import argparse
import base64
import re
from pathlib import Path

SCRIPT_ORDER = (
    "config.js",
    "i18n.js",
    "audio.js",
    "input.js",
    "model.js",
    "render.js",
    "app.js",
)


def build(root: Path, output: Path) -> None:
    index_path = root / "index.html"
    css_path = root / "styles.css"
    icon_path = root / "assets" / "icon.svg"
    if not index_path.is_file() or not css_path.is_file() or not icon_path.is_file():
        raise FileNotFoundError("index.html, styles.css, or assets/icon.svg is missing")

    html = index_path.read_text(encoding="utf-8")
    css = css_path.read_text(encoding="utf-8")
    icon_data = base64.b64encode(icon_path.read_bytes()).decode("ascii")

    html = re.sub(r'\s*<link rel="manifest"[^>]*>', "", html)
    html, count = re.subn(
        r'<link rel="icon"[^>]*>',
        f'<link rel="icon" href="data:image/svg+xml;base64,{icon_data}" type="image/svg+xml">',
        html,
        count=1,
    )
    if count != 1:
        raise RuntimeError("icon tag was not found exactly once")
    html, count = re.subn(
        r'<link rel="stylesheet" href="styles\.css">',
        f"<style>\n{css}\n</style>",
        html,
        count=1,
    )
    if count != 1:
        raise RuntimeError("stylesheet tag was not found exactly once")

    for filename in SCRIPT_ORDER:
        script_path = root / "src" / filename
        if not script_path.is_file():
            raise FileNotFoundError(script_path)
        source = script_path.read_text(encoding="utf-8")
        if "</script" in source.lower():
            raise ValueError(f"unsafe closing script token in {filename}")
        pattern = rf'<script(?: defer)? src="src/{re.escape(filename)}"></script>'
        html, count = re.subn(pattern, f"<script>\n{source}\n</script>", html, count=1)
        if count != 1:
            raise RuntimeError(f"script tag for {filename} was not found exactly once")

    standalone_flag = "<script>window.__PULSE_LINK_STANDALONE__ = true;</script>"
    html = html.replace("<head>", f"<head>\n  {standalone_flag}", 1)
    marker = "<!-- Standalone build: no external runtime assets, manifest, or service worker. -->"
    html = html.replace("</head>", f"  {marker}\n</head>", 1)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(html, encoding="utf-8")
    print(f"Built {output} ({output.stat().st_size:,} bytes)")


def main() -> None:
    default_root = Path(__file__).resolve().parent.parent
    parser = argparse.ArgumentParser()
    parser.add_argument("--root", type=Path, default=default_root)
    parser.add_argument(
        "--output",
        type=Path,
        default=default_root / "dist" / "pulse-link-overdrive-standalone.html",
    )
    args = parser.parse_args()
    build(args.root.resolve(), args.output.resolve())


if __name__ == "__main__":
    main()
